src/util/timelock.rs

```diff
@@ -0,0 +1,236 @@
+// Rust Bitcoin Library
+// Written in 2022 by
+//     Tobin C. Harding <me@tobin.cc>
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Bitcoin timelocks.
+//!
+//! Relative and absolute timelocks based on `nSequence` and `nLockTime`.
+//!
+
+#[cfg(feature = "std")]
+use std::error;
+
+use core::fmt;
+
+use blockdata::constants::{
+    LOCKTIME_THRESHOLD, SEQUENCE_LOCKTIME_DISABLE_FLAG, SEQUENCE_LOCKTIME_GRANULARITY,
+    SEQUENCE_LOCKTIME_TYPE_FLAG,
+};
+
+/// An absolute time lock, expires after either height or time.
+///
+/// This is nLockTime as defined in BIP-65.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Abs(u32);
+
+impl Abs {
+    /// Returns true if this absolute lock is lock-by-blockheight.
+    pub fn is_block_time(self) -> bool {
+        self.0 < LOCKTIME_THRESHOLD
+    }
+
+    /// Returns true if this absolute lock is lock-by-blocktime.
+    pub fn is_block_height(self) -> bool {
+        self.0 >= LOCKTIME_THRESHOLD
+    }
+
+    /// Returns true if this timelock is the same type as `other` timelock.
+    pub fn is_same_type(&self, other: Self) -> bool {
+        self.is_block_time() && other.is_block_time()
+            || self.is_block_height() && other.is_block_height()
+    }
+
+    /// Returns true if the timelock is locked at `age`.
```

Comment:
What `age` means is not described at all. I'm not familiar in detail with timelocks, maybe that's why. A sentence or two about it might help me and others like it.

Reply:
These docs are hard to write clearly, can you give feedback on this sentence please
```
/// Returns true if this timelock is expired at the specific blocktime/blockheight `age`.
```

Reply:
It is better and I understand it.

Though I would go with somethig:

```
`age` is a value corresponding to the type of the lock: either blocktime or blockheight
```

Also it made me think - there is kind no type safety here at all.

Reply:
Scratch some of that previous comments.

Looking at this now is  see that relative vs absolute timelocks are somewhat different things take from different parts (tx vs inputs) (yeah, I know my consensus-fu is not deep&fresh; but that's why I'm lurking here and it's a PoV closer to a noob that is learning).

So the block vs height are determined at runtime. Maybe it would be better to make both `Absolute` and `Relative` timelocks `enum Name { BlockBased(...), HeightBased(...) }`

And instead of using Absolute vs Relative, use the underlying naming.  

`NLockTime` and `TxInSequence` or something. Sure, one is absolute and the other relative, but that is not what they *are*, only their property. Using the proper names will force the user to learn them, avoids the weird naming, and might lead to better API idea in other functions.

This would be kind of a different approach. Instead of trying to come up with some abstract unified concept of a "timelock", we just encode in type system what things actually are represented in the consensus encoding.

Again, just learning myself and exploring the design space.

Reply:
I really disagree, we should stick with "absolute" and "relative" since these words are in common usage and they map nicely into Miniscript. Their encoding inside transactions are implementation details.

Reply:
The sequence number has more meanings than just timelocks. If we exposed the raw sequence number then we wouldn't have a useful timelock API, and if we interpreted it as a timelock then it wouldn't be a sequence number anymore.

Reply:
> The sequence number has more meanings than just timelocks. If we exposed the raw sequence number then we wouldn't have a useful timelock API, and if we interpreted it as a timelock then it wouldn't be a sequence number anymore.

Possibly.

Though `Sequence` can have all the calls that `Rel` has right now and it would work all the same. `Rel` and `Abs` are already different types, so it doesn't change much about functionality anywhere. If sequence has other properties, etc. then you just add them to `impl TxInSequence { ... }` and everything still works exactly. Trying to squeeze all code related to timelock handling into one module is kind of an artificial constrain, and forces user to do an additional step of converting from the sequence to the another type, just to be able to use the timelock API.

BTW. I was thinking that there are this separate types `Abs` vs `Rel` and there is bunch of duplicated methods on both, and not a single trait to bind them together in any way, which makes it impossible to write any code that would abstract away the difference between them. Maybe there is never really any reason to want to handle any of these types, ignoring which one is it exactly. That would mean that these types are different things, and just clumped together only because "they are kind of similar". Or maybe adding a trait would be a good idea.

Reply:
I could imagine:

```rust
struct TxIn {
   ...
   pub sequence: TxInSequence;
   ...
}

impl TimeLock for TxInSequence { ... }
```

Reply:
I think right now `rust-bitcoin` is really holding onto this approach that I would see in low level oriented C/C++ where "here is the type that is about encoding" (basically everything in `blockdata::block::Block`, using raw types everywhere), and "here are the separate types for higher level logic". 

From my PoV, that raw intermediate step is strictly unnecessary. You can take bytes and decode them right away into higher level types. No `u64`s etc. anywhere, except internally. One could still get them by some `.as_u64()` or whatnot, but by default you go bytes -> high level, nicely typed things.

Reply:
At a risk of making people dislike me, I think that since most devs come from heavy C++ experience, they tend to want to write "C++, but safer", instead of "Haskell, but as fast as C".

Reply:
This is not about C vs Haskell. It is about providing an API for timelocks, whose data is sourced from raw numeric fields, whose meanings (in Bitcoin's consensus code) are raw numeric fields.

Reply:
> sourced from raw numeric fields, whose meanings (in Bitcoin's consensus code) are raw numeric fields

Exactly. It's all just bytes on the wire. We don't even need `Abs` and `Rel` types there at all. Couple of constants and free standing functions taking `u32` arguments would do. :)

Reply:
Free standing functions would still achieve my declared aim of having all the timeock code in a single place to catch bugs/misuse.

Reply:
@dpc I appreciate that you are trying to satirize me but that is literally how these values are interpreted by Bitcoin's consensus code.

In some contexts there are further rules to interpret them as timelocks (which this module addresses). In other contexts there are rules regarding finality/non-finality (which is purely a policy thing and non-consensus). You can use these values as (signed) counters alongside OP_CSV/OP_CLTV in some contexts (e.g. eltoo). At one point in the past the Liquid functionaries used these values to sideload some pegout-processing state. Future softforks may interpret these values yet differently.

The point of this module is that the timelock case, which is both very common and consensus-critical, as well as absurdly edge-casey, would benefit from having some canonical logic inside this library.